rdtools/system_analysis.py

```diff
@@ -0,0 +1,587 @@
+
+    def aggregate(self, normalized, insolation):
+        '''
+        Returns insolation-weighted normalized PV energy and the associated aggregated insolation
+
+        Parameters
+        ----------
+        normalized: Pandas Times Series (numeric), normalized pv energy time series
+        insolation: Pandas Time Series (numeric), Time series of insolation associated
+                    with each normalize_energy point
+
+        Returns
+        -------
+        (aggregated, aggregated_insolation)
+        Tuple of Pandas Time Series consisting of insolation-weighted aggregated normalized PV energy
+        and the associated aggregated insolation.
+        '''
+        aggregated = aggregation.aggregation_insol(normalized, insolation, self.aggregation_freq)
+        aggregated_insolation = insolation.resample(self.aggregation_freq).sum()
+
+        return aggregated, aggregated_insolation
+
+    def yoy_degradation(self, aggregated, **kwargs):
+        '''
+        Perform year-on-year degradation analysis on insolation-weighted aggregated energy yield.
+
+        Parameters
+        ----------
+        aggregated: Pandas Time Series of insolation-weighted aggregated normalized PV energy
+        kwargs: passed to degradation.degradation_year_on_year()
+
+        Returns
+        -------
+        dict or year-on-year results with keys:
+            'p50_rd': The median year-on-year degradation rate (numeric)
+            'rd_confidence_interval': lower and upper bounds of degradation rate
+                                      confidence interval(list)
+            'calc_info': Dict of detailed results (see degradation.degradation_year_on_year() docs)
+        '''
+
+        yoy_rd, yoy_ci, yoy_info = degradation.degradation_year_on_year(aggregated, **kwargs)
+
+        yoy_results = {
+            'p50_rd': yoy_rd,
+            'rd_confidence_interval': yoy_ci,
+            'calc_info': yoy_info
+        }
+
+        return yoy_results
+
+    def srr_soiling(self, aggregated, aggregated_insolation, **kwargs):
+        '''
+        Performs stochastic rate and recovery soiling analysis.
+
+        Parameters
+        ---------
+        aggregated: Pandas Time Series of insolation-weighted aggregated normalized PV energy
+        aggregated_insolation: Pandas Time Series of insolation aggregated at same level as aggregated
+        kwargs: passed to soiling.soiling_srr()
+
+        Returns
+        -------
+        dict or year-on-year results with keys:
+            'p50_sratio': The median year-on-year insolation-weighted soiling ratio (numeric)
+            'sratio_confidence_interval': lower and upper bounds of insolation-weighted soiling ratio
+                                          confidence interval(list)
+            'calc_info': Dict of detailed results (see soiling.soiling_srr() docs)
+        '''
+        if aggregated.index.freq != 'D' or aggregated_insolation.index.freq != 'D':
```

Comment:
Technically other frequency strings may be valid here as well -- `'24h'`, `'1440T'` etc.  Could instead check the output of `pd.infer_freq(aggregated.index)` to be safe.

Reply:
Just found another option: `aggregated.index.freq != pd.tseries.offsets.Day()`.  Not sure how far back in pandas history it would work though.